### PR TITLE
Improve PI unique packet generation

### PR DESCRIPTION
**Pull request by safchain**
This aims to fix generation of unique packet ID when injecting from multiple nodes to one

### CODE REVIEW

**Comment by dvandra** [comment]
This is very useful to get a unique flow from multiple nodes to one destination.
But we need something to identify the source node per unique flow.
We need something either in flow or capture to identify source node per unique flow.

I am working on this and will soon update a patch.

File: cmd/injector/standalone.go [hunk 1]
```diff
@@ -46,25 +47,25 @@ var (
...
-	cmd.Flags().StringVarP(&srcMAC, "srcMAC", "", "", "source node MAC")
-	cmd.Flags().StringVarP(&dstMAC, "dstMAC", "", "", "destination node MAC")
-	cmd.Flags().Uint16VarP(&srcPort, "srcPort", "", 0, "source port for TCP packet")
-	cmd.Flags().Uint16VarP(&dstPort, "dstPort", "", 0, "destination port for TCP packet")
+	cmd.Flags().StringVarP(&srcIP, "src-ip", "", "", "source node IP")
+	cmd.Flags().StringVarP(&dstIP, "dst-ip", "", "", "destination node IP")
+	cmd.Flags().StringVarP(&srcMAC, "src-mac", "", "", "source node MAC")
+	cmd.Flags().StringVarP(&dstMAC, "dst-mac", "", "", "destination node MAC")
+	cmd.Flags().Uint16VarP(&srcPort, "src-port", "", 0, "source port for TCP packet")
+	cmd.Flags().Uint16VarP(&dstPort, "dst-port", "", 0, "destination port for TCP packet")
```
Comment (dvandra) [inline cmd/injector/standalone.go]: Is these changes needed?

Comment (safchain) [reply cmd/injector/standalone.go]: Not really but make command line more consistent, this doesn't break the API

**Comment by safchain** [comment]
Using the mode unique per node we should be able to identify the flows. Let say we take the srcPort of 50000 and we do have 5 source nodes then we know that the source ports will be between 50000 and 50005. We can use the same approach with the ICMP ID

File: cmd/injector/standalone.go (hunk 1 above)
Comment (dvandra) [reply cmd/injector/standalone.go]: +1

**Comment by safchain** [comment]
run skydive-compile-tests

**Comment by safchain** [comment]
run skydive-cdd-overview-tests
run skydive-compile-tests
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-functional-tests-backend-orientdb
run skydive-functional-tests-backend-elasticsearch
run skydive-functional-tests-backend-orientdb

**Comment by safchain** [comment]
run skydive-cdd-overview-tests

**Comment by safchain** [comment]
run skydive-cdd-overview-tests
run skydive-functional-tests-backend-orientdb
run skydive-k8s-tests

File: ondemand/client/client.go [hunk 2]
```diff
@@ -180,25 +185,28 @@ func (o *OnDemandClient) unregisterTask(node *graph.Node, resource types.Resourc
...
 	}
 
 	nps := map[graph.Identifier]nodeTask{}
-	for _, i := range nodes {
-		switch i.(type) {
+	for _, nr := range nrs {
+		if nodeID, host, ok := toRegister(nr); ok {
+			nps[nodeID] = nodeTask{nodeID, host, nr.Resource}
+		}
+		/*switch i.(type) {
```
Comment (lebauce) [inline ondemand/client/client.go]: Should be removed instead of commented ?

File: flow/ondemand/client/client.go [hunk 3]
```diff
@@ -69,15 +69,33 @@ func (h *onDemandFlowHandler) ResourceName() string {
...
 	query := capture.GremlinQuery
 	query += fmt.Sprintf(".Dedup().Has('Captures.ID', NEE('%s'))", resource.ID())
 	if capture.Type != "" && !common.CheckProbeCapabilities(capture.Type, common.MultipleOnSameNodeCapability) {
 		query += fmt.Sprintf(".Has('Captures.Type', NEE('%s'))", capture.Type)
 	}
 	query += h.nodeTypeQuery
-	return h.applyGremlinExpr(query)
+
+	if nodes := h.applyGremlinExpr(query); len(nodes) > 0 {
+		for _, i := range nodes {
```
Comment (lebauce) [inline flow/ondemand/client/client.go]: Could this be done in applyGremlinExpr ?

File: api/types/types.go [hunk 4]
```diff
@@ -237,6 +237,26 @@ func (n *NodeRule) Validate() error {
...
 }
 
+const (
+	// PIModeUniqPerNode use a unique packet identifier per source node
+	PIModeUniqPerNode = iota
+	// PIModeRandom use random packet identifier for each packet
+	PIModeRandom
+
+	// PiTypeICMP4 icmpv4 packet
+	PiTypeICMP4 = "icmp4"
```
Comment (lebauce) [inline api/types/types.go]: nit: s/PiTypeICMP4/PITypeICMP4/

File: scripts/gendecoder/main.go [hunk 5]
```diff
@@ -93,7 +93,7 @@ func resolveSymbol(pkg, symbol string) (types.Object, error) {
 
 func handleBasic(getter *getter, name, kind string) {
 	switch kind {
-	case "int8", "int16", "int32", "int64", "uint8", "uint16", "uint32", "uint64":
+	case "int", "int8", "int16", "int32", "int64", "uint8", "uint16", "uint32", "uint64":
```
Comment (lebauce) [inline scripts/gendecoder/main.go]: We should probably add uint too then

File: api/types/types.go [hunk 6]
```diff
@@ -256,7 +276,7 @@ type PacketInjection struct {
 	ICMPID           uint16 `yaml:"ICMPID"`
 	Count            uint64 `yaml:"Count"`
 	Interval         uint64 `yaml:"Interval"`
-	Increment        bool   `yaml:"Increment"`
+	Mode             int    `yaml:"Mode"`
```
Comment (lebauce) [inline api/types/types.go]: Why int and not int64 ?

File: cmd/injector/standalone.go [hunk 7]
```diff
@@ -46,25 +47,25 @@ var (
...
 	cmd.Flags().StringVarP(&packetType, "type", "", "icmp4", "packet type: icmp4, icmp6, tcp4, tcp6, udp4 and udp6")
 	cmd.Flags().StringVarP(&payload, "payload", "", "", "payload")
 	cmd.Flags().StringVar(&pcap, "pcap", "", "PCAP file")
 	cmd.Flags().Uint16VarP(&id, "id", "", 0, "ICMP identification")
-	cmd.Flags().BoolVarP(&increment, "increment", "", false, "increment ICMP id for each packet")
-	cmd.Flags().Int64VarP(&incrementPayload, "incrementPayload", "", 0, "increase payload for each packet")
+	cmd.Flags().IntVarP(&mode, "mode", "", types.PIModeUniqPerNode, "specify type of modification between injections")
```
Comment (lebauce) [inline cmd/injector/standalone.go]: This should probably be a string. Or at least give the possible values

File: api/types/types.go (hunk 6 above)
Comment (safchain) [reply api/types/types.go]: no specific reason, will change

File: flow/ondemand/client/client.go (hunk 3 above)
Comment (safchain) [reply flow/ondemand/client/client.go]: applyGremlinExpr is called somewhere else so I would prefer to keep this as it is

File: cmd/injector/standalone.go [hunk 8]
```diff
@@ -46,25 +47,25 @@ var (
...
+	cmd.Flags().StringVarP(&dstMAC, "dst-mac", "", "", "destination node MAC")
+	cmd.Flags().Uint16VarP(&srcPort, "src-port", "", 0, "source port for TCP packet")
+	cmd.Flags().Uint16VarP(&dstPort, "dst-port", "", 0, "destination port for TCP packet")
 	cmd.Flags().StringVarP(&packetType, "type", "", "icmp4", "packet type: icmp4, icmp6, tcp4, tcp6, udp4 and udp6")
 	cmd.Flags().StringVarP(&payload, "payload", "", "", "payload")
 	cmd.Flags().StringVar(&pcap, "pcap", "", "PCAP file")
 	cmd.Flags().Uint16VarP(&id, "id", "", 0, "ICMP identification")
-	cmd.Flags().BoolVarP(&increment, "increment", "", false, "increment ICMP id for each packet")
-	cmd.Flags().Int64VarP(&incrementPayload, "incrementPayload", "", 0, "increase payload for each packet")
+	cmd.Flags().StringVarP(&mode, "mode", "", "unique", "specify mode of packet generation, `unique` or `random`")
```
Comment (lebauce) [inline cmd/injector/standalone.go]: Would be nice to check if the specified value is either "unique" or "random"

**Comment by lebauce** [comment]
run skydive-functional-tests-backend-orientdb

File: contrib/python/api/skydive/packet_injector.py [hunk 9]
```diff
@@ -23,7 +23,7 @@ class PacketInjection(object):
     def __init__(self, uuid="", src="", dst="", srcip="", dstip="", srcmac="",
                  dstmac="", srcport=0, dstport=0, type="icmp4", payload="",
                  trackingid="", icmpid=0, count=1, interval=0,
-                 increment=False, starttime="", ttl=64):
+                 mode=0, starttime="", ttl=64):
```
Comment (lebauce) [inline contrib/python/api/skydive/packet_injector.py]: Should be mode="" now

File: api/types/types.go [hunk 10]
```diff
@@ -270,10 +290,21 @@ func (pi *PacketInjection) GetName() string {
...
+		PITypeTCP4:  true,
+		PITypeTCP6:  true,
+		PITypeUDP4:  true,
+		PITypeUDP6:  true,
+	}
 	if _, ok := allowedTypes[pi.Type]; !ok {
 		return errors.New("given type is not supported")
 	}
+
+	if pi.Mode != "" && pi.Mode != "unique" && pi.Mode != "random" {
```
Comment (lebauce) [inline api/types/types.go]: pi.Mode != PIModeUniqPerNode && pi.Mode != PIModeRandom

**Comment by safchain** [comment]
run skydive-cdd-overview-tests

**Comment by safchain** [comment]
run skydive-cdd-overview-tests
